HTTP clients blog post (#17)
whatyouhide/whatyouhide.github.io: content/posts/2023-07-24-breakdown-of-http-clients-in-elixir/index.md

extra:
  cover_image: cover-image.png
---

Elixir's ecosystem has quite a few **HTTP clients** at this point. But what's
the *best one*? In this post, I want to break down a bunch of the clients we
have available. I'll give an overview of the clients I personally like the most,
and I'll talk about which clients are the best choice in different use cases.

<!-- more -->

![Cover image of a futuristic-looking box ring with a metallic snake inside of it](cover-image.png)

<span class="image-caption">This is AI generated, just to be clear</span>

## All Your Clients Are Belong to Us

So, let's take a whirlwind tour of some HTTP clients available in Elixir. We'll talk about these:

  * [Mint](#mint)
  * [Finch](#finch)
  * [Req](#req)
  * [httpc](#httpc)
  * [Tesla](#tesla)

This is not a comprehensive list of all the Elixir HTTP clients, but rather a
list of clients that I think make sense in different situation. At the end of
this post, you'll also find a mention of other well-known clients.

### Mint

Let's start with [Mint][mint]. Mint is arguably the lowest-level HTTP client
we've got in Elixir. It's essentially a *wrapper* around a raw TCP or SSL

> socket. Its job is to make the socket **aware of the network protocol**.

Reviewer: For Mint, I would mention early on that it is a stateless / processless abstraction in that it doesn't imply any "architecture".

Author: 👍

```

Then, you'd use [`Mint.HTTP.request/5`](docs-mint-http-request) to send a
request.

```elixir
{:ok, conn, request_ref} = Mint.HTTP.request(conn, "GET", "/", [], "")
```

Sending a request is analogous to sending raw binary data on a `:gen_tcp` or
`:ssl` socket: it's *not blocking*. The call to `request/5` returns right away,
giving you a **request reference** back. The underlying socket will eventually
receive a response as an Erlang message. At that point, you can use
`Mint.HTTP.stream/2` to turn that message into something that makes sense in
HTTP.

```elixir
receive do
  message ->
    {:ok, conn, responses} = Mint.HTTP.stream(conn, message)
    IO.inspect(responses)
end
#=> [
#=>   {:status, #Reference<...>, 200},
#=>   {:headers, #Reference<...>, [{"connection", "keep-alive"}, ...},
#=>   {:data, #Reference<...>, "<!DOCTYPE html>..."},
#=>   {:done, #Reference<...>}
#=> ]
```

Mint supports HTTP/1.1 and HTTP/2 out of the box, as well as WebSocket through [mint_web_socket].

#### When to Use Mint

Generally, ***don't use Mint***. Seriously. You know I mean this advice, because
I'm one of the [two][gh-whatyouhide] [people][gh-ericmj] who maintain and

> originally created Mint itself! However, for most use cases, Mint is too low

Reviewer: ```suggestion
originally created Mint itself! For most use cases, Mint is too low
```

interested.

#### Bonus: Why Isn't Mint in the Elixir Standard Library?

That's a great question! When we introduced Mint back in 2019, we [posted about
it][mint-blog-post] on Elixir's website. Our original intention was to ship Mint
with Elixir's standard library. This is also one of the reasons why we wrote
Mint in Elixir, instead of Erlang. However, we then realized that it worked well
as a standalone library, and including it into the standard library would
increase the cost of maintaining the language as well as potentially slow down
the development of Mint itself.

That said, I think of Mint as the "standard-library HTTP client", that is, the
low-level client that you'd expect in the standard library of a language like
Elixir.

### Finch

[Finch][finch] is a client built on top of Mint. It serves an important job in
the "pyramid of abstractions" of HTTP clients listed in this post: **pooling**.
Finch provides pooling for Mint connections. Using Mint on its own means
implementing some sort of strategy to store and pool connections, which is what
Finch provides.

Finch is quite smart about its pooling. It uses [nimble_pool] when pooling

> HTTP/1.1 connections. The nimble_pool library is a tiny resource-pool

Reviewer: ```suggestion
HTTP/1.1 connections. The [`nimble_pool`](https://github.com/dashbitco/nimble_pool) library is a tiny resource-pool
```

Author: It's liked in the sentence before this one 🙃

implementation heavily focused on a small resource-usage footprint as well as on

> performance. Since HTTP/2 works quite differently from HTTP/1.1, especially when
> it comes to persistent connections and multiplexed requests, nimble_pool uses a
> completely different pooling strategy when working with HTTP/2 connections.

Reviewer: ```suggestion
performance. Since HTTP/2 works quite differently from HTTP/1.1 and the former
is capable of multiplexing requests, Finch uses a completely different strategy
for HTTP/2, without any pooling. All of this is transparent to users.
```

The API that Finch provides is still quite low-level, with manual request building and such:

```elixir
{:ok, _} = Finch.start_link(name: MyFinch)

Finch.build(:get, "https://hex.pm") |> Finch.request(MyFinch)
#=> {:ok, %Finch.Response{...}}
```

However, but the convenience of pooling and reconnections that Finch provides is
fantastic.

Okay, when to use Finch then? Personally, I think Finch is a fantastic library
whenever you have performance-sensitive applications where you're ready to
sacrifice some of the convenience provided by "higher-level" clients. It's also
great when you know you'll have to make a lot of requests to the same host,

> since you can specify dedicated connection pools per host.

Reviewer: ```suggestion
since you can specify dedicated connection pools per host. This is specially
useful when talking to internal APIs within a company.
```

### Req

[Req][req] is a relatively-new kid on the block when it comes to HTTP clients in
Elixir. It's one of my favorite Elixir libraries out there.

```elixir
Req.get!("https://api.github.com/repos/wojtekmach/req").body["description"]
#=> "Req is a batteries-included HTTP client for Elixir."
```

It's built on top of Finch, and it takes a quite "functional-programming"
approach to HTTP. What I mean by that is that Req revolves around a
[`Req.Request`][req-request-docs] *data structure*, which you manipulate to add
options, callbacks, headers, and more before making an HTTP request.

```elixir
req =

>   %Req.Request{method: :get, url: "https://github.com/..."}

Reviewer 2: ```suggestion
  Req.Request.new(method: :get, url: "https://github.com/...")
```

(I just added it. :) Cause if you forgot, which I totally just did not, wrap the url in a URI.parse, it will give a bad error message.)

Author: Ahaha nice catch. Done.

```elixir
Mix.install([
  {:req, "~> 0.3.0"}
])

Req.get!("https://andrealeopardi.com").headers
#=> [
#=>   {"connection", "keep-alive"},
#=>   ...
#=> ]
```

Req is also a great fit to use in your applications. It provides a ton of
features and plugins to use for things like encoding and decoding request
bodies, instrumentation, authentication, and so much more. I'll take a quote
straight from Finch's README here:

> Most developers will most likely prefer to use the fabulous HTTP client
> [Req][req] which takes advantage of Finch's pooling and provides an extremely
> friendly and pleasant to use API.

So, yeah.

### httpc

While Mint is the lowest-level HTTP client I know of, there's another client
worth mentioning alongside it: [`httpc`][httpc]. `httpc` ships with the Erlang
standard library, making it the **only** HTTP client in the ecosystem that
*doesn't require any additional dependencies*. This is so appealing! There are
cases where not having dependencies is a huge bonus. For example, if you're a
library author, being able to make HTTP requests without having to bring in
additional dependencies can be great, because those additional dependencies

> would trickle down (as transitive dependencies) to all users of your library.

Reviewer: You can give tailwind and esbuild libraries as examples of httpc is used once or twice to download executables and then it is never used again, so no additional dependencies are really worth it.

particular HTTP client. This is a big deal as far as I'm concerned, because I've
worked on projects with tens and tens of dependencies. In those projects, many
of those dependencies were a result of exactly this, that is, different
libraries using different HTTP clients (and all their transitive dependencies).

## Choosing the Right Client

I've tried to write a bit about when to use each client so far, but to recap,
this is my loose recommendation:

| Client         | When                                                                                                                                                            |
| :---------     | :-------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| *Mint*         | You need 100% control on connections and request lifecycle                                                                                                      |
| *Mint*         | You already have a process architecture, and don't want to introduce any more processes                                                                         |

> | *Mint*         | You're a library author and you want to force as few dependencies as possible on your users                                                                     |

Reviewer 3: This point is a bit confusing since this seems to describe `:httpc`. Or maybe add an extra differentiator, e.g. "but you do not want to use :httpc for safety/performance"?

Author: Great callout. Added!

> | *Finch*        | You need a low-level client with high performance, but you want pooling built in                                                                                |

Reviewer: ```suggestion
| *Finch*        | You need a low-level client with high performance, transparent HTTP/1 (with pooling) and HTTP/2 (with multiplexing) support      |
```

> | *Req*          | Scripting                                                                                                                                                       |

Reviewer 3: I know this is the opposite of your intention, but putting scripting first might give the impression that Req is for scripting before being for production apps?

Author: Good catch. Moved.

> | *Req*, *Tesla* | You want to build tailored HTTP clients in your application                                                                                                     |

Reviewer 3: "Tailored" seems quite "advanced", I'm afraid people who are just looking for a lib to make some HTTP calls might not recognize themselves.
